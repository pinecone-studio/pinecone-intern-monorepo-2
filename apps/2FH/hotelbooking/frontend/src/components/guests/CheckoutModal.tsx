'use client';
import React, { useState } from 'react';
import { useUpdateBookingMutation, BookingStatus, Response } from '@/generated';
import { Dialog, DialogContent, DialogHeader, DialogTitle, DialogTrigger } from '@/components/ui/dialog';
import { Button } from '@/components/ui/button';
import { Badge } from '@/components/ui/badge';
import { RadioGroup, RadioGroupItem } from '@/components/ui/radio-group';
import { Label } from '@/components/ui/label';
import { Loader2 } from 'lucide-react';

interface CheckoutModalProps {
  bookingId: string;
  currentStatus: string;
  onStatusUpdate?: () => void;
}

const CheckoutModal = ({ bookingId, currentStatus, onStatusUpdate }: CheckoutModalProps) => {
  const [open, setOpen] = useState(false);
  const [selectedStatus, setSelectedStatus] = useState(currentStatus);
  const [error, setError] = useState<string | null>(null);
  const [updateBooking, { loading }] = useUpdateBookingMutation();

  const statusOptions = [
    { value: BookingStatus.Booked, label: 'Booked', color: 'bg-blue-600 text-white' },
    { value: BookingStatus.Completed, label: 'Completed', color: 'bg-green-600 text-white' },
    { value: BookingStatus.Cancelled, label: 'Cancelled', color: 'bg-orange-500 text-white' },
  ];

  const handleStatusUpdate = async () => {
    try {
      setError(null);
      const result = await updateBooking({
        variables: {
          updateBookingId: bookingId,
          input: {
            status: selectedStatus as BookingStatus,
          },
        },
      });

      if (result.data?.updateBooking === Response.Success) {
        setOpen(false);
        onStatusUpdate?.();
      } else {
        setError('Failed to update booking status. Please try again.');
      }
    } catch (error) {
      console.error('Error updating booking status:', error);
      setError('An error occurred while updating the booking status. Please try again.');
    }
  };

  const handleOpenChange = (newOpen: boolean) => {
    setOpen(newOpen);
    if (newOpen) {
      setError(null);
    }
  };

  return (
    <Dialog open={open} onOpenChange={handleOpenChange}>
      <DialogTrigger asChild>
        <Button className="bg-blue-600 hover:bg-blue-700 text-white">Change Status</Button>
      </DialogTrigger>
      <DialogContent className="sm:max-w-md">
        <DialogHeader>
          <DialogTitle>Update Booking Status</DialogTitle>
        </DialogHeader>

        <div className="space-y-6">
          <div>
            <Label className="text-sm font-medium text-gray-600 mb-3 block">Current Status</Label>
            <Badge className={statusOptions.find((s) => s.value === currentStatus)?.color || 'bg-gray-600 text-white'}>{currentStatus.charAt(0) + currentStatus.slice(1).toLowerCase()}</Badge>
          </div>

          <div>
            <Label className="text-sm font-medium text-gray-600 mb-3 block">Select New Status</Label>
            <RadioGroup value={selectedStatus} onValueChange={setSelectedStatus} className="space-y-3">
              {statusOptions.map((status) => (
                <div key={status.value} className="flex items-center space-x-2">
                  <RadioGroupItem value={status.value} id={status.value} />
                  <Label htmlFor={status.value} className="flex items-center gap-2 cursor-pointer">
                    <Badge className={status.color}>{status.label}</Badge>
                  </Label>
                </div>
              ))}
            </RadioGroup>
          </div>

          {error && (
            <div className="bg-red-50 border border-red-200 rounded-md p-3">
              <p className="text-sm text-red-600">{error}</p>
            </div>
          )}

          <div className="flex justify-end gap-3">
            <Button variant="outline" onClick={() => setOpen(false)} disabled={loading}>
              Cancel
            </Button>
            <Button onClick={handleStatusUpdate} disabled={loading || selectedStatus === currentStatus} className="bg-blue-600 hover:bg-blue-700 text-white">
              {loading && <Loader2 className="mr-2 h-4 w-4 animate-spin" />}
              Update Status
            </Button>
          </div>
        </div>
      </DialogContent>
    </Dialog>
  );
};

export default CheckoutModal;
