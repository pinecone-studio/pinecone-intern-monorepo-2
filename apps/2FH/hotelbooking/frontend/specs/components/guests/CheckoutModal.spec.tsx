/* eslint-disable  */
import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import CheckoutModal from '@/components/guests/CheckoutModal';
import { useUpdateBookingMutation } from '@/generated';

// Mock dependencies
jest.mock('@/generated', () => ({
  useUpdateBookingMutation: jest.fn(),
}));

const mockUpdateBooking = jest.fn();
const mockUseUpdateBookingMutation = useUpdateBookingMutation as jest.MockedFunction<typeof useUpdateBookingMutation>;

describe('CheckoutModal', () => {
  const defaultProps = {
    bookingId: 'booking-1',
    currentStatus: 'BOOKED',
    onStatusUpdate: jest.fn(),
  };

  beforeEach(() => {
    jest.clearAllMocks();
    mockUseUpdateBookingMutation.mockReturnValue([mockUpdateBooking, { loading: false, error: undefined }] as any);
  });

  it('renders modal trigger button', () => {
    render(<CheckoutModal {...defaultProps} />);

    expect(screen.getByText('Change Status')).toBeInTheDocument();
  });

  it('opens modal when trigger button is clicked', () => {
    render(<CheckoutModal {...defaultProps} />);

    const triggerButton = screen.getByText('Change Status');
    fireEvent.click(triggerButton);

    expect(screen.getByText('Update Booking Status')).toBeInTheDocument();
  });

  it('displays current status', () => {
    render(<CheckoutModal {...defaultProps} />);

    const triggerButton = screen.getByText('Change Status');
    fireEvent.click(triggerButton);

    expect(screen.getByText('Current Status')).toBeInTheDocument();
    expect(screen.getAllByText('Booked')[0]).toBeInTheDocument();
  });

  it('displays all status options', () => {
    render(<CheckoutModal {...defaultProps} />);

    const triggerButton = screen.getByText('Change Status');
    fireEvent.click(triggerButton);

    expect(screen.getByText('Select New Status')).toBeInTheDocument();
    expect(screen.getAllByText('Booked')[0]).toBeInTheDocument();
    expect(screen.getAllByText('Completed')[0]).toBeInTheDocument();
    expect(screen.getAllByText('Cancelled')[0]).toBeInTheDocument();
  });

  it('handles status selection', () => {
    render(<CheckoutModal {...defaultProps} />);

    const triggerButton = screen.getByText('Change Status');
    fireEvent.click(triggerButton);

    const completedOption = screen.getByLabelText('Completed');
    fireEvent.click(completedOption);

    expect(completedOption).toBeChecked();
  });

  it('calls updateBooking mutation when update button is clicked', async () => {
    mockUpdateBooking.mockResolvedValue({
      data: { updateBooking: { id: 'booking-1' } },
    } as any);

    const onStatusUpdate = jest.fn();
    render(<CheckoutModal {...defaultProps} onStatusUpdate={onStatusUpdate} />);

    const triggerButton = screen.getByText('Change Status');
    fireEvent.click(triggerButton);

    const completedOption = screen.getByLabelText('Completed');
    fireEvent.click(completedOption);

    const updateButton = screen.getByText('Update Status');
    fireEvent.click(updateButton);

    await waitFor(() => {
      expect(mockUpdateBooking).toHaveBeenCalledWith({
        variables: {
          updateBookingId: 'booking-1',
          input: {
            status: 'COMPLETED',
          },
        },
      });
    });

    await waitFor(() => {
      expect(onStatusUpdate).toHaveBeenCalled();
    });
  });

  it('closes modal after successful update', async () => {
    mockUpdateBooking.mockResolvedValue({
      data: { updateBooking: { id: 'booking-1' } },
    } as any);

    render(<CheckoutModal {...defaultProps} />);

    const triggerButton = screen.getByText('Change Status');
    fireEvent.click(triggerButton);

    const completedOption = screen.getByLabelText('Completed');
    fireEvent.click(completedOption);

    const updateButton = screen.getByText('Update Status');
    fireEvent.click(updateButton);

    await waitFor(() => {
      expect(screen.queryByText('Update Booking Status')).not.toBeInTheDocument();
    });
  });

  it('handles update error', async () => {
    const consoleError = jest.spyOn(console, 'error').mockImplementation(() => {});
    mockUpdateBooking.mockRejectedValue(new Error('Update failed'));

    render(<CheckoutModal {...defaultProps} />);

    const triggerButton = screen.getByText('Change Status');
    fireEvent.click(triggerButton);

    const completedOption = screen.getByLabelText('Completed');
    fireEvent.click(completedOption);

    const updateButton = screen.getByText('Update Status');
    fireEvent.click(updateButton);

    await waitFor(() => {
      expect(consoleError).toHaveBeenCalledWith('Error updating booking status:', expect.any(Error));
    });

    consoleError.mockRestore();
  });

  it('disables update button when loading', () => {
    mockUseUpdateBookingMutation.mockReturnValue([mockUpdateBooking, { loading: true, error: undefined }] as any);

    render(<CheckoutModal {...defaultProps} />);

    const triggerButton = screen.getByText('Change Status');
    fireEvent.click(triggerButton);

    const updateButton = screen.getByText('Update Status');
    expect(updateButton).toBeDisabled();
  });

  it('disables update button when no status change is selected', () => {
    render(<CheckoutModal {...defaultProps} />);

    const triggerButton = screen.getByText('Change Status');
    fireEvent.click(triggerButton);

    const updateButton = screen.getByText('Update Status');
    expect(updateButton).toBeDisabled();
  });

  it('enables update button when different status is selected', () => {
    render(<CheckoutModal {...defaultProps} />);

    const triggerButton = screen.getByText('Change Status');
    fireEvent.click(triggerButton);

    const completedOption = screen.getByLabelText('Completed');
    fireEvent.click(completedOption);

    const updateButton = screen.getByText('Update Status');
    expect(updateButton).not.toBeDisabled();
  });

  it('closes modal when cancel button is clicked', () => {
    render(<CheckoutModal {...defaultProps} />);

    const triggerButton = screen.getByText('Change Status');
    fireEvent.click(triggerButton);

    const cancelButton = screen.getByText('Cancel');
    fireEvent.click(cancelButton);

    expect(screen.queryByText('Update Booking Status')).not.toBeInTheDocument();
  });

  it('disables cancel button when loading', () => {
    mockUseUpdateBookingMutation.mockReturnValue([mockUpdateBooking, { loading: true, error: undefined }] as any);

    render(<CheckoutModal {...defaultProps} />);

    const triggerButton = screen.getByText('Change Status');
    fireEvent.click(triggerButton);

    const cancelButton = screen.getByText('Cancel');
    expect(cancelButton).toBeDisabled();
  });

  it('shows loading spinner when updating', () => {
    mockUseUpdateBookingMutation.mockReturnValue([mockUpdateBooking, { loading: true, error: undefined }] as any);

    render(<CheckoutModal {...defaultProps} />);

    const triggerButton = screen.getByText('Change Status');
    fireEvent.click(triggerButton);

    const updateButton = screen.getByText('Update Status');
    expect(updateButton).toHaveTextContent('Update Status');
  });

  it('handles different current statuses', () => {
    const propsWithCompletedStatus = {
      ...defaultProps,
      currentStatus: 'COMPLETED',
    };

    render(<CheckoutModal {...propsWithCompletedStatus} />);

    const triggerButton = screen.getByText('Change Status');
    fireEvent.click(triggerButton);

    expect(screen.getAllByText('Completed')[0]).toBeInTheDocument();
  });

  it('handles cancelled status', () => {
    const propsWithCancelledStatus = {
      ...defaultProps,
      currentStatus: 'CANCELLED',
    };

    render(<CheckoutModal {...propsWithCancelledStatus} />);

    const triggerButton = screen.getByText('Change Status');
    fireEvent.click(triggerButton);

    expect(screen.getAllByText('Cancelled')[0]).toBeInTheDocument();
  });

  it('renders status badges with correct colors', () => {
    render(<CheckoutModal {...defaultProps} />);

    const triggerButton = screen.getByText('Change Status');
    fireEvent.click(triggerButton);

    // Get all "Booked" elements and select the first one (current status)
    const bookedBadges = screen.getAllByText('Booked');
    expect(bookedBadges[0]).toHaveClass('bg-blue-600', 'text-white');

    // Get all "Completed" elements and select the first one (radio option)
    const completedBadges = screen.getAllByText('Completed');
    expect(completedBadges[0]).toHaveClass('bg-green-600', 'text-white');

    // Get all "Cancelled" elements and select the first one (radio option)
    const cancelledBadges = screen.getAllByText('Cancelled');
    expect(cancelledBadges[0]).toHaveClass('bg-orange-500', 'text-white');
  });

  it('handles status update without onStatusUpdate callback', async () => {
    mockUpdateBooking.mockResolvedValue({
      data: { updateBooking: { id: 'booking-1' } },
    } as any);

    const propsWithoutCallback = {
      bookingId: 'booking-1',
      currentStatus: 'BOOKED',
    };

    render(<CheckoutModal {...propsWithoutCallback} />);

    const triggerButton = screen.getByText('Change Status');
    fireEvent.click(triggerButton);

    const completedOption = screen.getByLabelText('Completed');
    fireEvent.click(completedOption);

    const updateButton = screen.getByText('Update Status');
    fireEvent.click(updateButton);

    await waitFor(() => {
      expect(mockUpdateBooking).toHaveBeenCalledWith({
        variables: {
          updateBookingId: 'booking-1',
          input: {
            status: 'COMPLETED',
          },
        },
      });
    });
  });

  it('matches snapshot', () => {
    const { container } = render(<CheckoutModal {...defaultProps} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot when modal is open', () => {
    render(<CheckoutModal {...defaultProps} />);

    const triggerButton = screen.getByText('Change Status');
    fireEvent.click(triggerButton);

    const { container } = render(<CheckoutModal {...defaultProps} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('handles unknown status gracefully', () => {
    const propsWithUnknownStatus = {
      ...defaultProps,
      currentStatus: 'UNKNOWN_STATUS',
    };

    render(<CheckoutModal {...propsWithUnknownStatus} />);

    const triggerButton = screen.getByText('Change Status');
    fireEvent.click(triggerButton);

    // Should display the unknown status with fallback styling
    expect(screen.getByText('Current Status')).toBeInTheDocument();
    expect(screen.getByText('Unknown_status')).toBeInTheDocument(); // The status gets formatted
  });
});
