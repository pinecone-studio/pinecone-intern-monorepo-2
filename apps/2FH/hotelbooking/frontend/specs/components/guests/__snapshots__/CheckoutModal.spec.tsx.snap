// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckoutModal matches snapshot 1`] = `
<button
  aria-controls="radix-:r3f:"
  aria-expanded="false"
  aria-haspopup="dialog"
  class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 h-10 px-4 py-2 bg-blue-600 hover:bg-blue-700 text-white"
  data-state="closed"
  type="button"
>
  Change Status
</button>
`;

exports[`CheckoutModal matches snapshot when modal is open 1`] = `
<button
  aria-controls="radix-:r3o:"
  aria-expanded="false"
  aria-haspopup="dialog"
  class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 h-10 px-4 py-2 bg-blue-600 hover:bg-blue-700 text-white"
  data-state="closed"
  type="button"
>
  Change Status
</button>
`;
